libnetdata/dictionary/dictionary.c

```diff
@@ -253,78 +582,362 @@ int dictionary_del(DICTIONARY *dict, const char *name) {
+
+    dfe->dict = dict;
+    dfe->started_ut = now_realtime_usec();
+
+    dictionary_read_lock(dict);
+
+    NAME_VALUE *nv = dict->first_item;
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+void *dictionary_foreach_next(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return NULL;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(likely(nv)) {
+        // TODO: decrement reference counter on nv
+
+        // get the next item
+        nv = nv->next;
+    }
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv->next;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+usec_t dictionary_foreach_done(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return 0;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(nv) {
+        // TODO: decrement reference counter on nv
+        ;
+    }
+
+    dictionary_unlock((DICTIONARY *)dfe->dict);
+    dfe->dict = NULL;
+    dfe->last_position_index = NULL;
+    dfe->name = NULL;
+    dfe->value = NULL;
+
+    usec_t usec = now_realtime_usec() - dfe->started_ut;
+    dfe->started_ut = 0;
+
+    return usec;
+}
 
 // ----------------------------------------------------------------------------
 // API - walk through the dictionary
 // the dictionary is locked for reading while this happens
 // do not user other dictionary calls while walking the dictionary - deadlock!
 
-static int dictionary_walker(avl_t *a, int (*callback)(void *entry, void *data), void *data) {
-    int total = 0, ret = 0;
+int dictionary_walkthrough(DICTIONARY *dict, int (*callback)(const char *name, void *entry, void *data), void *data) {
+    dictionary_read_lock(dict);
+    int ret = linkedlist_namevalue_walkthrough_unsafe(dict, callback, data);
+    dictionary_unlock(dict);
+
+    return ret;
+}
+
+// ----------------------------------------------------------------------------
+// unit test
+
+static void dictionary_unittest_free_char_pp(char **pp, size_t entries) {
+    for(size_t i = 0; i < entries ;i++)
+        freez(pp[i]);
+
+    freez(pp);
+}
 
-    if(a->avl_link[0]) {
-        ret = dictionary_walker(a->avl_link[0], callback, data);
-        if(ret < 0) return ret;
-        total += ret;
+static char **dictionary_unittest_generate_names(size_t entries) {
+    char **names = mallocz(sizeof(char *) * entries);
+    for(size_t i = 0; i < entries ;i++) {
+        char buf[25 + 1] = "";
+        snprintfz(buf, 25, "name.%zu.0123456789.%zu \t !@#$%%^&*(),./[]{}\\|~`", i, entries / 2 + i);
+        names[i] = strdup(buf);
```

Comment:
We are using `freez` to clean the characters, but we are using `strdup` instead `strdupz`. I suggest to use our internal library here, because `freez` does math with the allocated address, and `strdup` is not implemented equally in all platforms we support.

Reply:
nice catch

Reply:
fixed it